manim/utils/decorators.py

```diff
@@ -0,0 +1,454 @@
+"""Function decorators."""
+
+__all__ = ["deprecated", "deprecated_params"]
+
+
+import re
+from typing import Any, Callable, Iterable, List, Optional, Tuple, Union
+
+from decorator import decorate, decorator
+
+from .. import logger
+
+
+def _get_callable_info(callable: Callable) -> Tuple[str, str]:
+    """Returns type and name of a callable.
+
+    Parameters
+    ----------
+    callable
+        The callable
+
+    Returns
+    -------
+    Tuple[str, str]
+        The type and name of the callable. Type can can be one of "class", "method" (for
+        functions defined in classes) or "function"). For methods, name is Class.method.
+    """
+    what = type(callable).__name__
+    name = callable.__qualname__
+    if what == "function" and "." in name:
+        what = "method"
+    elif what == "type":
+        what = "class"
+    return (what, name)
+
+
+def _deprecation_text_component(
+    since: Optional[str], until: Optional[str], message: str
+) -> str:
+    """Generates a text component used in deprecation messages.
+
+    Parameters
+    ----------
+    since
+        The version or date since deprecation
+    until
+        The version or date until removal of the deprecated callable
+    message
+        The reason for why the callable has been deprecated
+
+    Returns
+    -------
+    str
+        The deprecation message text component.
+    """
+    since = "" if since is None else f"since {since} "
+    until = (
+        "may be deleted in a future version"
+        if until is None
+        else f"is expected to be deleted after {until}"
+    )
+    return f"deprecated {since}and {until}. {message}"
```

Comment:
Isn't a space missing here ?

Reply:
If `since` is empty, then it will output an empty string, for which there is already a space preceding it. Otherwise, if `since` contains a version number or date, the output will insert a space after it.

Reply:
Oh ok